File: library/kotlin/src/io/envoyproxy/envoymobile/PulseClientImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import io.envoyproxy.envoymobile.engine.types.HistogramUnit

/**
 * Envoy implementation of `PulseClient`.
 */
internal class PulseClientImpl constructor(
  internal val engine: EnvoyEngine
) : PulseClient {

  override fun counter(vararg elements: Element): Counter {
    return CounterImpl(engine, elements.asList())
  }

  override fun gauge(vararg elements: Element): Gauge {
    return GaugeImpl(engine, elements.asList())
  }

  override fun histogram(unitMeasure: HistogramUnit, vararg elements: Element): Histogram {
    return HistogramImpl(engine, elements.asList(), unitMeasure)
  }
}

File: library/swift/PulseClientImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// Envoy implementation of PulseClient.
final class PulseClientImpl: NSObject {
  private let engine: EnvoyEngine

  init(engine: EnvoyEngine) {
    self.engine = engine
    super.init()
  }
}

extension PulseClientImpl: PulseClient {
  func counter(elements: [Element]) -> Counter {
    return CounterImpl(elements: elements, tags: [], engine: self.engine)
  }

  func counter(elements: [Element], tags: [Tag]) -> Counter {
    return CounterImpl(elements: elements, tags: tags, engine: self.engine)
  }

  func gauge(elements: [Element]) -> Gauge {
    return GaugeImpl(elements: elements, tags: [], engine: self.engine)
  }

  func gauge(elements: [Element], tags: [Tag]) -> Gauge {
    return GaugeImpl(elements: elements, tags: tags, engine: self.engine)
  }

  func timer(elements: [Element]) -> Timer {
    return TimerImpl(elements: elements, tags: [], engine: self.engine)
  }

  func timer(elements: [Element], tags: [Tag]) -> Timer {
    return TimerImpl(elements: elements, tags: tags, engine: self.engine)
  }

  func distribution(elements: [Element], tags: [Tag]) -> Distribution {
    return DistributionImpl(elements: elements, tags: tags, engine: self.engine)
  }

  func distribution(elements: [Element]) -> Distribution {
    return DistributionImpl(elements: elements, tags: [], engine: self.engine)
  }
}
